File: app/src/main/java/org/digitalcampus/oppia/utils/custom_prefs/TimePreference.java
package org.digitalcampus.oppia.utils.custom_prefs;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;

import androidx.preference.DialogPreference;

public class TimePreference extends DialogPreference {
    public int hour = 0;
    public int minute = 0;

    public static int parseHour(String value) {
        try {
            String[] time = value.split(":");
            return (Integer.parseInt(time[0]));
        } catch (Exception e) {
            return 0;
        }
    }

    public static int parseMinute(String value) {
        try {
            String[] time = value.split(":");
            return (Integer.parseInt(time[1]));
        } catch (Exception e) {
            return 0;
        }
    }

    public static String timeToString(int h, int m) {
        return String.format("%02d", h) + ":" + String.format("%02d", m);
    }

    public TimePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected Object onGetDefaultValue(TypedArray a, int index) {
        return a.getString(index);
    }


    @Override
    protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {
        String value;
        if (restoreValue) {
            if (defaultValue == null) value = getPersistedString("00:00");
            else value = getPersistedString(defaultValue.toString());
        } else {
            value = defaultValue.toString();
        }

        hour = parseHour(value);
        minute = parseMinute(value);
    }

    public void setValue(String value) {
        hour = parseHour(value);
        minute = parseMinute(value);
        persistString(value);
    }

    public String getValue() {
        return timeToString(hour, minute);
    }


    public void persistStringValue(String value) {
        persistString(value);
    }
}

File: app/src/main/java/org/digitalcampus/oppia/utils/custom_prefs/TimePreference.kt
package org.digitalcampus.oppia.utils.custom_prefs

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import androidx.preference.DialogPreference

open class TimePreference(context: Context, attrs: AttributeSet) : DialogPreference(context, attrs) {
    @JvmField
    var hour = 0
    @JvmField
    var minute = 0
    override fun onGetDefaultValue(a: TypedArray, index: Int): Any? {
        return a.getString(index)
    }

    override fun onSetInitialValue(restoreValue: Boolean, defaultValue: Any) {
        val value: String = if (restoreValue) {
            getPersistedString(defaultValue.toString())
        } else {
            defaultValue.toString()
        }
        hour = parseHour(value)
        minute = parseMinute(value)
    }

    var value: String
        get() = timeToString(hour, minute)
        set(value) {
            hour = parseHour(value)
            minute = parseMinute(value)
            persistString(value)
        }

    fun persistStringValue(value: String?) {
        persistString(value)
    }

    companion object {
        fun parseHour(value: String): Int {
            return try {
                val time = value.split(":".toRegex()).toTypedArray()
                time[0].toInt()
            } catch (e: Exception) {
                0
            }
        }

        fun parseMinute(value: String): Int {
            return try {
                val time = value.split(":".toRegex()).toTypedArray()
                time[1].toInt()
            } catch (e: Exception) {
                0
            }
        }

        fun timeToString(h: Int, m: Int): String {
            return String.format("%02d", h) + ":" + String.format("%02d", m)
        }
    }
}